client/src/main/java/com/netflix/conductor/client/automator/TaskPollExecutor.java

```diff
@@ -0,0 +1,246 @@
+            task = MetricsContainer.getPollTimer(taskType)
+                .record(() -> taskClient.pollTask(taskType, worker.getIdentity(), domain));
+
+            if (Objects.nonNull(task) && StringUtils.isNotBlank(task.getTaskId())) {
+                MetricsContainer.incrementTaskPollCount(taskType, 1);
+                LOGGER.debug("Polled task: {} of type: {} in domain: '{}', from worker: {}",
+                    task.getTaskId(), taskType, domain, worker.getIdentity());
+
+                CompletableFuture<Task> taskCompletableFuture = CompletableFuture.supplyAsync(() ->
+                    processTask(task, worker), executorService);
+
+                taskCompletableFuture.whenComplete(this::finalizeTask);
+            } else {
+                // no task was returned in the poll, release the permit
+                pollingSemaphore.complete();
+            }
+        } catch (Exception e) {
+            // release the permit if exception is thrown during polling, because the thread would not be busy
+            pollingSemaphore.complete();
+            MetricsContainer.incrementTaskPollErrorCount(worker.getTaskDefName(), e);
+            LOGGER.error("Error when polling for tasks", e);
+        }
+    }
+
+    void shutdown() {
+        shutdownExecutorService(executorService);
+    }
+
+    void shutdownExecutorService(ExecutorService executorService) {
+        int timeout = 10;
+        try {
+            if (executorService.awaitTermination(timeout, TimeUnit.SECONDS)) {
+                LOGGER.debug("tasks completed, shutting down");
+            } else {
+                LOGGER.warn(String.format("forcing shutdown after waiting for %s second", timeout));
+                executorService.shutdownNow();
+            }
+        } catch (InterruptedException ie) {
+            LOGGER.warn("shutdown interrupted, invoking shutdownNow");
+            executorService.shutdownNow();
+            Thread.currentThread().interrupt();
+        }
+    }
+
+    @SuppressWarnings("FieldCanBeLocal")
+    private final Thread.UncaughtExceptionHandler uncaughtExceptionHandler = (thread, error) -> {
+        // JVM may be in unstable state, try to send metrics then exit
+        MetricsContainer.incrementUncaughtExceptionCount();
+        LOGGER.error("Uncaught exception. Thread {} will exit now", thread, error);
+    };
+
+
+    private Task processTask(Task task, Worker worker) {
+        LOGGER.debug("Executing task: {} of type: {} in worker: {} at {}", task.getTaskId(), task.getTaskDefName(),
+            worker.getClass().getSimpleName(), worker.getIdentity());
+        try {
+            executeTask(worker, task);
+        } catch (Throwable t) {
+            task.setStatus(Task.Status.FAILED);
+            TaskResult result = new TaskResult(task);
+            handleException(t, result, worker, task);
+        } finally {
+            pollingSemaphore.complete();
```

Comment:
Apart from this finally block, I don't see `processTask` adding any value to `executeTask`. Why not just use `executeTask` and get rid of this?

Reply:
`executeTask` calls execute on the worker and handles uncaught exceptions. Then it tries to  `updateTask` with the server.
`processTask` is only responsible for handling exceptions during this update. Combining different exception cases based on the flow within the same method would be confusing.